### PR TITLE
Refactor library types to use ItemType + GADTs

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

**Comment by phadej** [comment]
- Principled way of `partitionItems` (vs. previous `partitionEithers`)
- Removal of `CriteriaAndData` and `CriteriaAndData` types
- Both would help adding third type of item
    - Note how more straight forward tests are, no need for `allBookSortCriteriaAndStart` -like values

File: library-app/src/Futurice/App/Library/Pages/IndexPage.hs [hunk 1]
```diff
@@ -12,85 +14,86 @@ import Futurice.App.Library.API
...
-            a_ [class_ (currentPage Book <> "button"), href_ $ linkToText $ fieldLink indexPageGet (Just $ BookSort SortTitle) Nothing Nothing Nothing Nothing Nothing Nothing Nothing] $ "Books"
-            a_ [class_ (currentPage BoardGame <> "button"), href_ $ linkToText $ fieldLink indexPageGet (Just $ BoardGameSort SortName) Nothing Nothing Nothing Nothing Nothing Nothing Nothing] $ "Boardgames"
-    case cd of
-      BookCD bookCriteria books -> do
-          let bookIndexPageLink = indexPageLink (BookSort bookCriteria)
-          paginationLinks cd
+            a_ [class_ (currentPage Book      <> "button"), recordHref_ indexPageGet (Just $ MkSome $ BookSort SortTitle)     Nothing Nothing Nothing Nothing Nothing Nothing Nothing] $ "Books"
+            a_ [class_ (currentPage BoardGame <> "button"), recordHref_ indexPageGet (Just $ MkSome $ BoardGameSort SortName) Nothing Nothing Nothing Nothing Nothing Nothing Nothing] $ "Boardgames"
+    case crit of
+      BookSort bookCriteria | books <- map fromItemBook itemInfos -> do
```
Comment (neurogoo) [inline library-app/src/Futurice/App/Library/Pages/IndexPage.hs]: How does this map know that ItemInfo is type `ItemInfo 'Book`?

File: library-app/src/Futurice/App/Library/Types/ItemType.hs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+-- | Fancy 'partitionEithers'.
+partitionItems :: HasItemType f => (a -> Some f) -> [a] -> ItemBuckets f
+partitionItems g = go where
+    go []       = ItemBuckets [] []
+    go (x : xs) = case g x of
+        MkSome f -> case itemType f of
+            SBook      -> ItemBuckets (f : fs0) fs1
+            SBoardGame -> ItemBuckets fs0 (f : fs1)
+      where
+        ~(ItemBuckets fs0 fs1) = go xs
```
Comment (neurogoo) [inline library-app/src/Futurice/App/Library/Types/ItemType.hs]: What does this syntax do?

File: library-app/src/Futurice/App/Library/Pages/IndexPage.hs (hunk 1 above)
Comment (phadej) [reply library-app/src/Futurice/App/Library/Pages/IndexPage.hs]: when we match on `crit`, we learn that `ty` is `'Book`.

File: library-app/src/Futurice/App/Library/Types/ItemType.hs (hunk 2 above)
Comment (phadej) [reply library-app/src/Futurice/App/Library/Types/ItemType.hs]: it's a lazy match. https://wiki.haskell.org/Lazy_pattern_match

See e.g. http://hackage.haskell.org/package/base-4.12.0.0/docs/src/Data.Either.html#partitionEithers

**Pull request merged**